types/rpc-handler.ts

```diff
@@ -60,26 +61,113 @@ export class RPCHandler implements HandlerInterface {
     this.getNetworkName.bind(this);
     this.getNetworkRpcs.bind(this);
     this.testRpcPerformance.bind(this);
+    this.consensusCall.bind(this);
+    this.getFirstAvailableRpcProvider.bind(this);
+    this._rpcService = new RPCService(this);
   }
 
   /**
-   * Loops through all RPCs for a given network id and returns a provider with the first successful network.
+   * @NOTICE # This method is intended for read operations only.
+   * Performing write operations with this method is dangerous as it is executed and confirmed by potentially all supported network providers.
+   *
+   * @DEV `quorumThreshold` is a template literal to enforce at the type level that the value is a decimal
+   * between 0 and 1 without the need for external packages or custom classes which burden the user and codebase.
+   *
+   * @DESCRIPTION Similar to `Ethers.FallbackProvider`, this method validates the response from multiple nodes and returns true if the response is consistent across all nodes above the quorum threshold.
+   *
+   * @EXAMPLE
+   * - `consensusCall({ method: "eth_blockNumber", params: [] }, "0.5")`
+   * - `consensusCall({ method: "eth_getTransactionByHash", params: ["0x1234"] }, "0.8")`
    */
-  public async getFirstAvailableRpcProvider() {
-    const rpcList = [...networkRpcs[this._networkId].rpcs];
-    shuffleArray(rpcList);
-    for (const rpc of rpcList) {
+  public async consensusCall<TMethodReturnData = unknown>(requestPayload: RequestPayload, quorumThreshold: `0.${number}`): Promise<TMethodReturnData> {
+    if (this._runtimeRpcs.length === 0) {
+      await this.testRpcPerformance();
+    }
+
+    if (this._runtimeRpcs.length === 1) {
+      throw new Error("Only one RPC available, could not reach consensus");
+    }
+
+    const quorum = Math.ceil(this._runtimeRpcs.length * Number(quorumThreshold));
+    const rpcs = this._runtimeRpcs;
+    const results: PromiseResult[] = [];
+
+    for (const rpc of rpcs) {
       try {
-        const result = await RPCService.makeRpcRequest(rpc.url, this._rpcTimeout, { "Content-Type": "application/json" });
+        const result = await this._rpcService.makeRpcRequest(requestPayload, { rpcUrl: rpc, rpcTimeout: this._rpcTimeout });
         if (result.success) {
-          return new JsonRpcProvider({ url: rpc.url, skipFetchSetup: true }, Number(this._networkId));
-        } else {
-          console.error(`Failed to reach endpoint ${rpc.url}. ${result.error}`);
+          results.push(result);
         }
       } catch (err) {
-        console.error(`Failed to reach endpoint ${rpc.url}. ${err}`);
+        this.log("error", `Failed to reach endpoint ${rpc}.\n ${String(err)}`);
       }
     }
+
+    if (results.length < quorum) {
+      throw new Error(`Failed to reach consensus of ${quorum} with ${results.length} successful RPC responses`);
+    }
+
+    const rpcResults = results.map((res) => res.data?.result);
+
+    const matchingResults = rpcResults.reduce(
```

Comment:
I don't understand this code block. The point is that we should query RPC nodes and check that at least `quorum` returns the same result. Does it work this way?

Reply:
Yeah it increments each time a response is the same. 

In this logic we were matching the hash, but as you've highlighted that needs changed to something more generic.